ledger/participant-integration-api/src/main/scala/platform/store/state/CachingContractsReader.scala

```diff
@@ -0,0 +1,244 @@
+        headsToBeSignaled.add((offset, eventSequentialId, enqueuedAt))
+        ()
+      }
+  }
+
+  def consumeFrom(implicit
+      loggingContext: LoggingContext
+  ): Flow[ContractStateEvent, Unit, NotUsed] =
+    Flow[ContractStateEvent]
+      .map {
+        case el @ ContractStateEvent.Created(
+              contractId,
+              _,
+              globalKey,
+              _,
+              eventOffset,
+              eventSequentialId,
+            ) =>
+          logger.debug(
+            s"State events update: Created(contractId=$contractId, globalKey=$globalKey, offset=$eventOffset, eventSequentialId=$eventSequentialId"
+          )
+          el
+        case el @ ContractStateEvent.Archived(
+              contractId,
+              _,
+              eventOffset,
+              eventSequentialId,
+            ) =>
+          logger.debug(
+            s"State events update: Archived(contractId=$contractId, offset=$eventOffset, eventSequentialId=$eventSequentialId"
+          )
+          el
+        case el @ LedgerEndMarker(eventOffset, eventSequentialId) =>
+          logger.debug(
+            s"Ledger end reached: $eventOffset -> $eventSequentialId "
+          )
+          el
+      }
+      .map {
+        case ContractStateEvent.Created(
+              contractId,
+              contract,
+              globalKey,
+              flatEventWitnesses,
+              _,
+              eventSequentialId,
+            ) =>
+          globalKey.foreach(
+            keyCache.put(_, eventSequentialId, Assigned(contractId, flatEventWitnesses))
+          )
+          contractsCache.put(contractId, eventSequentialId, Active(contract, flatEventWitnesses))
+          eventSequentialId
+        case ContractStateEvent.Archived(
+              contractId,
+              stakeholders,
+              _,
+              eventSequentialId,
+            ) =>
+          contractsCache.put(
+            contractId,
+            eventSequentialId,
+            Archived(eventSequentialId, stakeholders),
+          )
+          eventSequentialId
+        case other => other.eventSequentialId // Just pass the seq id downstream
+      }
+      .map(idx => {
+        cacheIndex.set(idx)
+        Option(headsToBeSignaled.peek())
+          .foreach { oldestHeadInQueue =>
+            logger.debug(s"New cache index $idx vs oldest head in queue $oldestHeadInQueue")
+            headsToBeSignaled.removeIf { case dekd @ (offset, seqId, enqueuedAt) =>
+              if (seqId <= idx) {
+                logger.debug(s"Dequeued $dekd and signaling new ledger end")
+                metrics.daml.index.cacheCatchup
+                  .update(System.nanoTime() - enqueuedAt, TimeUnit.NANOSECONDS)
+                signalGlobalNewLedgerEnd(offset)
+                true
+              } else false
+            }
+            ()
+          }
+
+        metrics.daml.indexer.currentStateCacheSequentialIdGauge.updateValue(idx)
+      })
```

Comment:
Minor: I would break this long flow into multiple subflows, e.g.:
```suggestion
    def consumeFrom(implicit
      loggingContext: LoggingContext
  ): Flow[ContractStateEvent, Unit, NotUsed] =
    Flow[ContractStateEvent]
      .via(debugFlow)
      .via(updateCache)
      .map(_.eventSequentialId)
      .map(idx => {
```

```
  def updateCache(implicit
                  loggingContext: LoggingContext
                 ): Flow[ContractStateEvent, ContractStateEvent, NotUsed] =
    Flow[ContractStateEvent].map {

```

```
  def debugFlow(implicit
                  loggingContext: LoggingContext
                 ): Flow[ContractStateEvent, ContractStateEvent, NotUsed] = {
    Flow[ContractStateEvent].map {
```